causalml/inference/meta/xlearner.py

```diff
@@ -570,21 +606,42 @@ def fit(self, X, treatment, y):
             self.models_tau_c[group].fit(X_filt[w == 0], d_c)
             self.models_tau_t[group].fit(X_filt[w == 1], d_t)
 
-    def predict(self, X, p, treatment=None, y=None, return_components=False, verbose=True):
+    def predict(self, X, treatment=None, y=None, p=None, return_components=False,
+                return_p_score=False, verbose=True):
         """Predict treatment effects.
 
         Args:
             X (np.matrix or np.array or pd.Dataframe): a feature matrix
-            p (np.ndarray or pd.Series or dict): an array of propensity scores of float (0,1) in the single-treatment
-                case; or, a dictionary of treatment groups that map to propensity vectors of float (0,1)
             treatment (np.array or pd.Series, optional): a treatment vector
             y (np.array or pd.Series, optional): an outcome vector
-
+            p (np.ndarray or pd.Series or dict, optional): an array of propensity scores of float (0,1) in the
+                single-treatment case; or, a dictionary of treatment groups that map to propensity vectors of
+                float (0,1); if None will run ElasticNetPropensityModel() to generate the propensity scores.
+            return_components (bool, optional): whether to return outcome for treatment and control seperately
+            return_p_score (bool, optional): whether to return propensity score
+            verbose (bool, optional): whether to output progress logs
         Returns:
             (numpy.ndarray): Predictions of treatment effects.
         """
         X, treatment, y = convert_pd_to_np(X, treatment, y)
-        check_p_conditions(p, self.t_groups)
+
+        if p is None:
+            logger.info('Generating propensity score')
+            p = dict()
+            p_model = dict()
+            for group in self.t_groups:
+                mask = (treatment == group) | (treatment == self.control_name)
+                treatment_filt = treatment[mask]
+                X_filt = X[mask]
+                w_filt = (treatment_filt == group).astype(int)
+                w = (treatment == group).astype(int)
+                p[group], p_model[group] = compute_propensity_score(X=X_filt, treatment=w_filt,
+                                                                    X_pred=X, treatment_pred=w)
+            self.propensity_model = p_model
+            self.propensity = p
+        else:
+            check_p_conditions(p, self.t_groups)
```

Comment:
I think that we should not train a propensity model in `predict()` but use either propensity scores provided or propensity model(s) trained during `fit()` instead.

Then, we can make sure that we're using the same propensity model(s) for `fit()` and `predict()`. Also we don't need to return propensity scores in `predict()` as well.

Reply:
Hi @jeongyoonlee, thanks for reviewing it! For X-Learner, we didn't need `propensity_score`  in `fit()` and only needed when calculating the predictions of treatment effects which we have in `predict()` step.  Our original `fit()` function didn't require `p`, so I didn't add there.  For R-Learner, we did what you suggested here train a propensity model in `fit()` and return the propensity score and models(s).

Reply:
I see. Then, let's set `self.propensity` and `self.propensity_model` as in R learner without returning `p` in `predict()`.

Reply:
Thanks! Changed accordingly.